Adds optax update guide. (#1774)
google/flax: docs/howtos/optax_update_guide.rst

Optax Update Guide
==================

> Code samples below are executable in

Author: Yeah, we have a couple of these Colabs.

But in this case I care more about the side-by-side of the `flax.optim` and `optax` code to make users' lives easier.

But I thought it's still very useful to have exactly the same code available in a Colab that runs end to end.

`Colab <https://colab.research.google.com/github/google/flax/blob/main/docs/notebooks/optax_update_guide.ipynb>`_.

.. testsetup::

  import flax
  import jax
  import jax.numpy as jnp
  import flax.linen as nn
  import optax

  batch = {'image': jnp.ones([1, 28, 28, 1]), 'label': jnp.array([0])}
  ds_train = [batch]

>   get_ds_train = lambda: [batch]

Reviewer: nit: do we need both lines? (they seem to be the same thing)

Author: Added a comment to clarify.

  loss = lambda params, batch: jnp.array(0.)

Replacing ``flax.optim`` with ``optax``
---------------------------------------

Optax has drop-in replacements for all of Flax's optimizers. Refer to Optax's
documentation `Common Optimizers <https://optax.readthedocs.io/en/latest/api.html>`_
for API details.

The usage is very similar, with the difference that ``optax`` does not keep a
copy of the ``params``, so they need to be passed around separately. Flax
provides the utility :py:class:`~flax.training.train_state.TrainState` to store
optimizer state, parameters, and other associated data in a single dataclass.

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  @jax.jit
  def train_step(optimizer, batch):
    grads = jax.grad(loss)(optimizer.target, batch)
    return optimizer.apply_gradient(grads)

  optimizer_def = flax.optim.Momentum(
      learning_rate, momentum)
  optimizer = optimizer_def.create(variables['params'])

  for batch in get_ds_train():
    optimizer = train_step(optimizer, batch)

  ---

  @jax.jit
  def train_step(params, opt_state, batch):
    grads = jax.grad(loss)(params, batch)

>     updates, opt_state = tx.update(grads, opt_state)

Author: Moved above.

    params = optax.apply_updates(params, updates)
    return params, opt_state

>   tx = optax.sgd(learning_rate, momentum)

Author: Ah, interesting. I originally took the `tx` from the Optax README (under "Gradient Transformations"), and didn't realize that `optimizer` would be the more common name for the chain of transformations..

But then I started using that in our FLIP and all our examples. So I think for consistency I'll rather keep `tx`.

      optax.trace(decay=momentum),
      # 2. Step: multiply result from step 1 with negative learning rate.
      # Note that `optax.apply_updates()` simply adds the final updates to the
      # parameters, so we must make sure to flip the sign here for gradient
      # descent.
      optax.scale(-learning_rate),
  )

Weight Decay
------------

Some of Flax's optimizers also include a weight decay. In Optax, the weight
decay can be added as another "gradient transformation"
|optax.add_decayed_weights()|_ that adds an update derived from the parameters.

.. |optax.add_decayed_weights()| replace:: ``optax.add_decayed_weights()``
.. _optax.add_decayed_weights(): https://optax.readthedocs.io/en/latest/api.html#optax.add_decayed_weights

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  optimizer_def = flax.optim.Adam(
      learning_rate, weight_decay=weight_decay)
  optimizer = optimizer_def.create(variables['params'])

  ---

  tx = optax.chain(
      optax.scale_by_adam(),
      optax.add_decayed_weights(weight_decay),
      # params -= learning_rate * (adam(grads) + params * weight_decay)
      optax.scale(-learning_rate),

>   )

Reviewer: nit: optax defines `optax.adamw()` for this common case (https://github.com/deepmind/optax/blob/master/optax/_src/alias.py#L273)

Author: Added a comment.

For learning rate schedules, Flax allows to overwrite hyper parameters when
applying the gradients. Optax keeps a step counter and provides this as an
argument to a function for scaling the updates added with
|optax.scale_by_schedule()|_. Optax also allows to specify functions to
inject arbitrary scalar values for other gradient updates via
|optax.inject_hyperparams()|_.

Read more about learning rate schedules in the :doc:`lr_schedule` guide.

.. |optax.scale_by_schedule()| replace:: ``optax.scale_by_schedule()``
.. _optax.scale_by_schedule(): https://optax.readthedocs.io/en/latest/api.html#optax.scale_by_schedule
.. |optax.inject_hyperparams()| replace:: ``optax.inject_hyperparams()``
.. _optax.inject_hyperparams(): https://optax.readthedocs.io/en/latest/api.html#optax.inject_hyperparams

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  def train_step(step, optimizer, batch):
    grads = jax.grad(loss)(optimizer.target, batch)
    return step + 1, optimizer.apply_gradient(grads, learning_rate=schedule(step))

  ---

  tx = optax.chain(
      optax.trace(decay=momentum),
      # Note that we still want a negative value for scaling the updates!

>       optax.scale_by_schedule(lambda step: -schedule(step)),

Reviewer: it could be worth mentioning that optax has a number of standard schedules predefined: https://optax.readthedocs.io/en/latest/optax-101.html#weight-decay-schedules-and-clipping

Reviewer: i.e. you can do:

```python
schedule = optax.warmup_cosine_decay_schedule(
  init_value=0.0,
  peak_value=1.0,
  warmup_steps=50,
  decay_steps=1_000,
  end_value=0.0,
)

optimizer = optax.adamw(learning_rate=schedule)
```

Author: Good point, added a reference to the lr schedules, and mentioned that `learning_rate` of standard optimizers can be scalar or schedule.

> Optax Update Guide

Reviewer 2: As it reads now it sounds like it's about updating Optax.

I propose renaming to "Upgrading my codebase to Optax"

==================

> 

Reviewer 2: I think a short intro is in order here, e.g. "The built-in `flax.optim` package is being deprecated, and we instead recommend all users use the [Optax](link to github)] library. Follow this guide to upgrade your codebase."

Code samples below are executable in

> `Colab <https://colab.research.google.com/github/google/flax/blob/main/docs/notebooks/optax_update_guide.ipynb>`_.

Reviewer 2: What's the user journey where someone wants to run this Colab? (as opposed to copying code from the RTD page). I ask because I'd rather avoid having to remember to update two places if we change a variable name in this doc, etc.

Reviewer 2: Or, another version of the question: Should we have colabs for *all* of our HOWTOs? If so, maybe we should find a way to automatically convert them. Or is this one special? I'm hesitant to introduce this new expectation by adding just one guide with a Colab, unless we want to go in that direction more broadly.

Reviewer 2: (But no need to block merging this PR on this discussion, just wanted to bring it up. We can resolve separately)

Author: I think the user journey is to provide users with a working example that they can use to experiment with different gradient transformations to replicate their original setup. Of course they could do this on their own starting with the code blocks from the transition guide, but there's still some more code involved and it's annoying having to copy this together and a nicer experience to start with a working Colab.

I first tried to have a Colab that is automatically synced, but then realized two problems:

1. I think it's very useful to have the side-by-side in the update guide. This would require us to create a new Sphinx plugin that knows how to take two cells in the Colab, do some pre-processing, and then put them into a diff html table.
2. The Colab will necessarily contain some different code. For example, I want to provide a more realistic example that can be used int he Colab to play around with optimizer settings in a meaningful way. In RST on the other hand we want to have a cheap setup that does not load anything from `tdfs`. This would probably be a simpler Sphinx plugin replacing some code block with another code block when transcribing the Colab into RST.

If we have a Sphinx plugin implementing the 2 features mentioned above then I'll gladly use it. But it's too much work to implement, and I'd rather have to sync the two documents for the added functionality. I don't expect us to update this code frequently, so it's more about remembering (and even if we forgot to update a small thing, it wouldn't be too bad).

Author: Ack. Leaving comment open for better visiblity.

|optax.inject_hyperparams()|_.

Read more about learning rate schedules in the :doc:`lr_schedule` guide.

Read more about schedules defined in Optax under `Optimizer Schedules
<https://optax.readthedocs.io/en/latest/api.html#optimizer-schedules>`_. the
standard optimizers (like ``optax.adam()``, ``optax.sgd()`` etc.) also accept a
learning rate schedule as a parameter for `learning_rate`.

.. |optax.scale_by_schedule()| replace:: ``optax.scale_by_schedule()``
.. _optax.scale_by_schedule(): https://optax.readthedocs.io/en/latest/api.html#optax.scale_by_schedule
.. |optax.inject_hyperparams()| replace:: ``optax.inject_hyperparams()``
.. _optax.inject_hyperparams(): https://optax.readthedocs.io/en/latest/api.html#optax.inject_hyperparams

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  def train_step(step, optimizer, batch):
    grads = jax.grad(loss)(optimizer.target, batch)
    return step + 1, optimizer.apply_gradient(grads, learning_rate=schedule(step))

  ---

  tx = optax.chain(
      optax.trace(decay=momentum),
      # Note that we still want a negative value for scaling the updates!
      optax.scale_by_schedule(lambda step: -schedule(step)),
  )

> Multiple Optimizers

Reviewer 2: ```suggestion
Multiple Optimizers or only updating a subset of parameters
```

(Since `optax.masked` can be used for that as well)

-------------------

> Flax's :py:class:`flax.optim.MultiOptimizer` can be used to compose multiple

Reviewer 2: ```suggestion
In Flax, Traversals are used to specify which parameters should be updated by an optimizer. And you can combine Traversals using :py:class:`flax.optim.MultiOptimizer` to apply different optimizers on different parameters. The equivalent in Optax is `optax.masked()` and `optax.chain()`.
```

.. testsetup::

  import flax
  import jax
  import jax.numpy as jnp
  import flax.linen as nn
  import optax

  # Note: this is the minimal code required to make below code run. See in the
  # Colab linked above for a more meaningful definition of datasets etc. 
  batch = {'image': jnp.ones([1, 28, 28, 1]), 'label': jnp.array([0])}
  ds_train = [batch]
  get_ds_train = lambda: [batch]
  model = nn.Dense(1)
  variables = model.init(jax.random.PRNGKey(0), batch['image'])
  learning_rate, momentum, weight_decay, grad_clip_norm = .1, .9, 1e-3, 1.
  loss = lambda params, batch: jnp.array(0.)

Replacing ``flax.optim`` with ``optax``
---------------------------------------

Optax has drop-in replacements for all of Flax's optimizers. Refer to Optax's
documentation `Common Optimizers <https://optax.readthedocs.io/en/latest/api.html>`_
for API details.

The usage is very similar, with the difference that ``optax`` does not keep a
copy of the ``params``, so they need to be passed around separately. Flax
provides the utility :py:class:`~flax.training.train_state.TrainState` to store
optimizer state, parameters, and other associated data in a single dataclass.

> .. codediff::

Reviewer 3: Very cool you used the side-by-side view, great to see that it is actually useful 😄 . Not sure if you are aware of this, but if you add a "`#!`" to the end of a line, that line will be highlighted. (I can also see that it may not be necessary in this HOWTO, so feel free to ignore.)

Author: Ah, I was wondering why no lines were highlighted (I somehow assumed it would be automatic).

Good to know, but I went through the diffs (or rather "side by sides") and didn't really find value in aligning the code snippets and highlighting the changes (there are too few common lines). But good to know for my next HOTWO.

  :title_left: ``flax.optim``
  :title_right: ``optax``

  @jax.jit

>   def train_step(optimizer, batch):

Reviewer 3: Have you considered aligning this side-by-side view nicer? RIght now the two `train_step` functions are at different places, which makes it a bit less easy to compare.

Author: Thinking again about this, I'll undo an earlier change proposed by Ross to align things even more nicely. I think in this case it's more useful to have code aligned than define variables in global scope before defining the function uses them (see Ross's comment above for the motivation to rearrange in the first place)

Reviewer 3: Ah sorry about that, I didn't realize Ross already commented on this and I should have checked. Ok I'm fine either way!

    grads = jax.grad(loss)(optimizer.target, batch)
    return optimizer.apply_gradient(grads)

  optimizer_def = flax.optim.Momentum(
      learning_rate, momentum)
  optimizer = optimizer_def.create(variables['params'])

  for batch in get_ds_train():
    optimizer = train_step(optimizer, batch)

  ---

  tx = optax.sgd(learning_rate, momentum)
  params = variables['params']
  opt_state = tx.init(params)

  @jax.jit
  def train_step(params, opt_state, batch):
    grads = jax.grad(loss)(params, batch)
    updates, opt_state = tx.update(grads, opt_state)
    params = optax.apply_updates(params, updates)
    return params, opt_state

  for batch in ds_train:
    params, opt_state = train_step(params, opt_state, batch)

Composable Gradient Transformations
-----------------------------------

> Note though that |optax.sgd()|_ is simply a wrapper for the sequential

Reviewer 3: "Note though" seems a bit of a strange way to start a new section. Maybe just say "`|optax.sgd()|_` is simply a wrapper for the..."?

Author: Done.

application of two gradient transformations. Instead of using this alias, it is
common to use |optax.chain()|_ to combine multiple of these generic building
blocks.

.. |optax.sgd()| replace:: ``optax.sgd()``
.. _optax.sgd(): https://optax.readthedocs.io/en/latest/api.html#optax.sgd
.. |optax.chain()| replace:: ``optax.chain()``
.. _optax.chain(): https://optax.readthedocs.io/en/latest/api.html#chain

.. codediff::
  :title_left: Pre-defined alias
  :title_right: Combining transformations

  # Note that the aliases follow the convention to use positive
  # values for the learning rate by default.
  tx = optax.sgd(learning_rate, momentum)

  ---

  tx = optax.chain(
      # 1. Step: keep a trace of past updates and add to gradients. 
      optax.trace(decay=momentum),
      # 2. Step: multiply result from step 1 with negative learning rate.
      # Note that `optax.apply_updates()` simply adds the final updates to the
      # parameters, so we must make sure to flip the sign here for gradient
      # descent.
      optax.scale(-learning_rate),
  )

Weight Decay
------------

> Some of Flax's optimizers also include a weight decay. In Optax, the weight

Reviewer 3: weight decay can even be added using a "decay mask", you could mention that since the Flax BERT finetuning example in HuggingFace uses it: https://github.com/huggingface/transformers/blob/master/examples/flax/text-classification/run_flax_glue.py#L189 (maybe just point to it as a reference, so users can look there if they want to see an example?)

Author: Added reference to `optax.adamw()`. Also created https://github.com/deepmind/optax/pull/284 to clarify the usage.

decay can be added as another "gradient transformation"
|optax.add_decayed_weights()|_ that adds an update derived from the parameters.

.. |optax.add_decayed_weights()| replace:: ``optax.add_decayed_weights()``
.. _optax.add_decayed_weights(): https://optax.readthedocs.io/en/latest/api.html#optax.add_decayed_weights

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  optimizer_def = flax.optim.Adam(
      learning_rate, weight_decay=weight_decay)
  optimizer = optimizer_def.create(variables['params'])

  ---

  # (Note that you could also use `optax.adamw()` in this case)
  tx = optax.chain(
      optax.scale_by_adam(),
      optax.add_decayed_weights(weight_decay),
      # params -= learning_rate * (adam(grads) + params * weight_decay)
      optax.scale(-learning_rate),
  )
  # Note that you'll need to specify `params` when computing the udpates:
  # tx.update(grads, opt_state, params)

> Gradient Clipping

Reviewer 3: Great example!

-----------------

Training can be stabilized by clipping gradients to a global norm (`Pascanu et
al, 2012 <https://arxiv.org/abs/1211.5063>`_). In Flax this is often done by
processing the gradients before passing them to the optimizer. With Optax this
becomes just another gradient transformation |optax.clip_by_global_norm()|_.

.. |optax.clip_by_global_norm()| replace:: ``optax.clip_by_global_norm()``
.. _optax.clip_by_global_norm(): https://optax.readthedocs.io/en/latest/api.html#optax.clip_by_global_norm

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  def train_step(optimizer, batch):
    grads = jax.grad(loss)(optimizer.target, batch)
    grads_flat, _ = jax.tree_flatten(grads)
    global_l2 = jnp.sqrt(sum([jnp.vdot(p, p) for p in grads_flat]))
    g_factor = jnp.minimum(1.0, grad_clip_norm / global_l2)
    grads = jax.tree_map(lambda g: g * g_factor, grads)
    return optimizer.apply_gradient(grads)

  ---

  tx = optax.chain(
      optax.clip_by_global_norm(grad_clip_norm),
      optax.trace(decay=momentum),
      optax.scale(-learning_rate),
  )

Learning Rate Schedules
-----------------------

> For learning rate schedules, Flax allows to overwrite hyper parameters when

Reviewer 3: ```suggestion
For learning rate schedules, Flax allows overwriting hyperparameters when
```

Author: Done.

> applying the gradients. Optax keeps a step counter and provides this as an

Reviewer 3: ```suggestion
applying the gradients. Optax maintains a step counter and provides this as an
```

Author: Done.

argument to a function for scaling the updates added with

> |optax.scale_by_schedule()|_. Optax also allows to specify functions to

Reviewer 3: ```suggestion
|optax.scale_by_schedule()|_. Optax also allows specifying functions to
```

Author: Done.

inject arbitrary scalar values for other gradient updates via
|optax.inject_hyperparams()|_.

Read more about learning rate schedules in the :doc:`lr_schedule` guide.

Read more about schedules defined in Optax under `Optimizer Schedules
<https://optax.readthedocs.io/en/latest/api.html#optimizer-schedules>`_. the
standard optimizers (like ``optax.adam()``, ``optax.sgd()`` etc.) also accept a

> learning rate schedule as a parameter for `learning_rate`.

Reviewer 3: ```suggestion
learning rate schedule as a parameter for ``learning_rate``.
```

Author: Linkified.

.. |optax.scale_by_schedule()| replace:: ``optax.scale_by_schedule()``
.. _optax.scale_by_schedule(): https://optax.readthedocs.io/en/latest/api.html#optax.scale_by_schedule
.. |optax.inject_hyperparams()| replace:: ``optax.inject_hyperparams()``
.. _optax.inject_hyperparams(): https://optax.readthedocs.io/en/latest/api.html#optax.inject_hyperparams

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  def train_step(step, optimizer, batch):
    grads = jax.grad(loss)(optimizer.target, batch)

>     return step + 1, optimizer.apply_gradient(grads, learning_rate=schedule(step))

Reviewer 3: You could consider making this line less wide since right now the side by side has a scrollbar on laptops, like this?

```suggestion
    grad = optimizer.apply_gradient(grads, learning_rate=schedule(step))
    return step+1, grad
```

Author: Note that it's the `optimizer`.

Also, we already use the same pattern in the other snippets, so I'd leave it as is for consistency.

  ---

  tx = optax.chain(
      optax.trace(decay=momentum),
      # Note that we still want a negative value for scaling the updates!
      optax.scale_by_schedule(lambda step: -schedule(step)),
  )

Multiple Optimizers or only updating a subset of parameters
-------------------

> In Flax, Traversals are used to specify which parameters should be updated by an optimizer. And you can combine Traversals using :py:class:`flax.optim.MultiOptimizer` to apply different optimizers on different parameters. The equivalent in Optax is `optax.masked()` and `optax.chain()`.

Reviewer 3: ```suggestion
In Flax, Traversals are used to specify which parameters should be updated by an optimizer. And you can combine Traversals using :py:class:`flax.optim.MultiOptimizer` to apply different optimizers on different parameters. The equivalent in Optax is ``optax.masked()`` and ``optax.chain()``.
```

Author: Updated + linkified.

> optimizers and only apply them to a part of the params pytree. Optax provides

Reviewer 3: This sentence doesn't seem to logically follow the previous one, maybe something went wrong?

Author: I accepted a change from Avital and didn't proof read. Cleaned up.

an efficient masking mechanism to achieve the same goal.

> Note that below example is using :py:mod:`flax.traverse_util` to create the

Reviewer 3: ```suggestion
Note that example below is using :py:mod:`flax.traverse_util` to create the
```

Author: Done.

boolean masks required by |optax.masked()|_ - alternatively you could also
create them manually, or use |optax.multi_transform()|_ that takes a
multivalent pytree to specify gradient transformations.

Beware that |optax.masked()|_ flattens the pytree internally and the inner
gradient transformations will only be called with that partial flattened view of
the params/gradients. This is not a problem usually, but it makes it hard to
nest multiple levels of masked gradient transformations (because the inner
masks will expect the mask to be defined in terms of the partial flattened view
that is not readily available outside the outer mask).

.. |optax.masked()| replace:: ``optax.masked()``
.. _optax.masked(): https://optax.readthedocs.io/en/latest/api.html#optax.masked
.. |optax.multi_transform()| replace:: ``optax.multi_transform()``
.. _optax.multi_transform(): https://optax.readthedocs.io/en/latest/api.html#optax.multi_transform

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

>   kernels = flax.traverse_util.ModelParamTraversal(lambda p, _: 'kernel' in p)
>   biases = flax.traverse_util.ModelParamTraversal(lambda p, _: 'bias' in p)
>   kernel_opt = flax.optim.Momentum(learning_rate, momentum)
>   bias_opt = flax.optim.Momentum(learning_rate * 0.1, momentum)
>   optimizer = flax.optim.MultiOptimizer(
>       (kernels, kernel_opt),
>       (biases, bias_opt)
>   ).create(variables['params'])

Reviewer 3: Will this align nicer with the RHS?

```suggestion
  kernels = flax.traverse_util.ModelParamTraversal(lambda p, _: 'kernel' in p)
  biases = flax.traverse_util.ModelParamTraversal(lambda p, _: 'bias' in p)
  
  kernel_opt = flax.optim.Momentum(learning_rate, momentum)
  bias_opt = flax.optim.Momentum(learning_rate * 0.1, momentum)
  
  
  optimizer = flax.optim.MultiOptimizer(
      (kernels, kernel_opt),
      (biases, bias_opt)
  ).create(variables['params'])
```

Author: Done.

gradient transformations will only be called with that partial flattened view of
the params/gradients. This is not a problem usually, but it makes it hard to
nest multiple levels of masked gradient transformations (because the inner
masks will expect the mask to be defined in terms of the partial flattened view
that is not readily available outside the outer mask).

.. |optax.masked()| replace:: ``optax.masked()``
.. _optax.masked(): https://optax.readthedocs.io/en/latest/api.html#optax.masked
.. |optax.multi_transform()| replace:: ``optax.multi_transform()``
.. _optax.multi_transform(): https://optax.readthedocs.io/en/latest/api.html#optax.multi_transform

.. codediff::
  :title_left: ``flax.optim``
  :title_right: ``optax``

  kernels = flax.traverse_util.ModelParamTraversal(lambda p, _: 'kernel' in p)
  biases = flax.traverse_util.ModelParamTraversal(lambda p, _: 'bias' in p)

  kernel_opt = flax.optim.Momentum(learning_rate, momentum)
  bias_opt = flax.optim.Momentum(learning_rate * 0.1, momentum)

  optimizer = flax.optim.MultiOptimizer(
      (kernels, kernel_opt),
      (biases, bias_opt)
  ).create(variables['params'])

  ---

>   kernels = flax.traverse_util.ModelParamTraversal(lambda p, _: 'kernel' in p)

Reviewer 2: Oh, I'm sorry to have missed this in the first round of review but I'm curious -- why do we need to be using `ModelParamTraversal` at all? Presumably other users of Optax, e.g. those using Haiku, have some other pattern? I'd love to eventually deprecate `ModelParamTraversal` as I don't think it's needed anymore, and if needed replace it with something pretty much like a tree_map that gives you the full path of subkeys rather than just the values.

Author: There's plenty of different ways how you can create those masks.

I took `ModelParamTraversal` because it's the most straightforward transition from existing code, but feel free to update the example to use `traverse_utils.flatten_dict()` instead.

Reviewer 2: OK I'll give it a shot.
